README.md
Pippel
============

[![Melpa](https://melpa.org/packages/pippel-badge.svg)](http://melpa.org/#/pippel)

This package is an Emacs frontend for the Python package manager pip. As pippel also uses `tabulated-list-mode`, it provides a similiar package menu like `package-list-packages`.

![](https://raw.githubusercontent.com/arifer612/pippel/master/pippel-menu.png)

## Usage

To use it call `M-x pippel-list-packages`.

Shortcuts for `pippel-package-menu-mode` buffers:

 * `m`     `pippel-menu-mark-unmark` remove mark
 * `d`     `pippel-menu-mark-delete` mark for deletion
 * `U`     `pippel-menu-mark-all-upgrades` mark all upgradable
 * `u`     `pippel-menu-mark-upgrade` mark for upgrade
 * `r`     `pippel-list-packages` refresh package list
 * `i`     `pippel-install-package` prompt user for packages
 * `x`     `pippel-menu-execute` perform marked package menu actions
 * `RET`   `pippel-menu-visit-homepage` follow link


## Installation

The package can be installed from MELPA.

    (require 'pippel)

## Changelog
- v1.4 <2021-12-05>
  + Fix for pip >21.1.3 (#c3f10e6)
  + Formmating using black and add docstrings (#0e62405)

- v1.3 <2021-06-15>
  + Fix for pip >19.1.1 (#2480fd3)
  + Maintainer changed to arifer612


CHANGELOG.org
#+TITLE: Changelog
#+DESCRIPTION: Records of changes made to Pippel

** v1.5 (alpha)
*** Additions:
+ CHANGELOG file
+ Unit test scripts and README file build script

*** Changes:
+ Access user library when not in virtualenv. Close #19 (#705715e)
+ Package organisation
  
** v1.4.1 [2022-04-17]

*** Additions:
+ The use of the `traceback` library for more verbose error debugging.
  
*** Fixes:
+ I/O logging error because of direct use of `main()` for InstallCommand and
  UninstallCommand.


** v1.4 [2021-12-05]
*** Changes:
+ Formatting using black and add docstrings. (#0e62405)
*** Fixes:
+ Breaking changes for pip >21.1.3. (#c3f10e6)

** v1.3 [2021-06-15]
*** Changes:
+ Maintainer changed to [[https://github.com/arifer612][arifer612]]
*** Fixes:
+ Breaking changes for pip >19.1.1. (#2480fd3)


test/test_pippel.py
"""Unit tests for pippel.py"""
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../")))  # noqa: E501,BLK100

from src.pippel import Server, PipBackend  # noqa: E402

server = PipBackend()


def test_list():
    """
    Verifies that the get_installed_packages command works.
    """
    res = server.get_installed_packages("test")
    return 0 if res else 1


def test_install():
    """
    Verifies that the install_package command works.
    """
    return server.install_package("zipp", "test")


def test_remove():
    """
    Verifies that the remove_package command works.
    """
    return server.remove_package("zipp", "test")


src/pippel.py
"""
Script for Pippel to probe Pip for installed packages and modify them.
Instructions and functions are to be executed through pippel.el, a detailed
documentation is provided in it.
"""

import json
import os
import traceback
import sys
import subprocess
from typing import Dict, List, Optional, Tuple

# Parse current Pip version
from pip import __version__ as pip_version

pip_version = tuple(map(int, pip_version.split(".")))
if len(pip_version) == 2:
    pip_version = (*pip_version, 0)


def compare_version(version_a, version_b):
    # type: (tuple, tuple) -> bool
    """
    Compares version tuples.
    Returns True if VERSION_A is greater than or equal to VERSION_B.
    """
    for rank_a, rank_b in zip(version_a, version_b):
        if rank_a != rank_b:
            return rank_a > rank_b
    return True

# Module imports depend on Pip version.
# pylint: disable=import-error, no-name-in-module
from pip._internal.utils.compat import stdlib_pkgs
if compare_version((19, 1, 1), pip_version):
    from pip.commands.install import InstallCommand
    from pip.commands.list import ListCommand
    from pip.commands.show import search_packages_info
    from pip.commands.uninstall import UninstallCommand
    from pip.utils import get_installed_distributions
else:
    from pip._internal.commands.install import InstallCommand
    from pip._internal.commands.list import ListCommand
    from pip._internal.commands.show import search_packages_info
    from pip._internal.commands.uninstall import UninstallCommand

    if compare_version(pip_version, (21, 2, 0)):
        # These are required to get the packages from 21.2
        from pip._internal.metadata import get_environment
        from typing import cast
    if not compare_version(pip_version, (21, 2, 3)):
        # get_installed_distributions is removed from 21.3
        from pip._internal.utils.misc import get_installed_distributions

# pylint: enable=import-error, no-name-in-module


class Server(object):
    """
    Server to communicate between Emacs and Pip.
    """

    def __init__(self, _stdin=None, _stdout=None):
        self.stdin = _stdin if _stdin else sys.stdin
        self.stdout = _stdout if _stdout else sys.stdout

    def write_out(self, payload, test_call):
        # type: (str, bool) -> None
        """
        Dumps PAYLOAD into STDOUT if TEST_CALL is False.
        """
        if not test_call:
            self.stdout.write(payload + "\n")

    def write_json(self, result, test_call):
        # type: (list, bool) -> None
        """
        Dumps result into STDOUT.
        """
        if not test_call:
            self.stdout.write(json.dumps(result) + "\n")
            self.stdout.flush()

    def read_json(self):
        # type: () -> Dict[str, str]
        """
        Parses STDIN as a dictionary.
        """
        line = self.stdin.readline()
        if not line:
            raise EOFError()
        return json.loads(line)

    def handle_request(self):
        # type: () -> None
        """
        Sends requests to server.
        """
        request = self.read_json()
        _method = request["method"]
        arg1 = request.get("arg1")
        arg2 = request.get("arg2")
        try:
            self.stdout.write("pip version: %s\n" % str(pip_version))
            self.stdout.write("Executing: %s(%s, %s)\n" %
                              (_method, arg1, arg2))
            method = getattr(self, _method)
            method(arg1, arg2)
            self.stdout.write("Pip finished\n")
        except Exception as exception:
            self.stdout.write("PIPPEL_ERROR <<<\n")
            self.stdout.write("%s\n\n" % exception)
            self.stdout.write("%s\n" % traceback.format_exc())
            self.stdout.write("PIPPEL_ERROR >>>\n")

    def serve_forever(self):
        """
        Keeps server alive until it breaks.
        """
        while True:
            try:
                self.handle_request()
            except (KeyboardInterrupt, EOFError, SystemExit):
                break


class PipBackend(Server):
    """
    Backend server for Pippel with methods.
    """

    @staticmethod
    def in_virtual_env():
        # type: () -> bool
        """
        Returns TRUE if package is run while in a virtual environment.
        """
        if hasattr(sys, "base_prefix"):
            return sys.base_prefix != sys.prefix
        if hasattr(sys, "real_prefix"):
            return sys.real_prefix != sys.prefix
        return False

    def get_installed_packages(self, params="", *args, **kwargs):
        # type: (str, tuple, dict) -> List[Dict[str, str]]
        """
        Retrieves all the installed packages in current environment as a dictionary.
        If USER is TRUE and the current python environment is not a virtual environment,
        only the user installed packages are retrieved.
        """
        if compare_version((19, 1, 0), pip_version):
            get_list = ListCommand()
        else:
            get_list = ListCommand("Pippel", "Backend server for the Pippel service.")

        if "test" in params:
            params = params.replace("test", "")
            test_call = True
        else:
            test_call = False

        if self.in_virtual_env() or not params:
            params = ""


        options, _ = get_list.parse_args(["--outdated", params])
        skip = set(stdlib_pkgs)

        if compare_version((21, 2, 0), pip_version):
            packages = [
                package
                for package in get_installed_distributions(
                        local_only=options.local,
                        user_only=options.user,
                        editables_only=options.editable,
                        include_editables=options.include_editable,
                        paths=options.path,
                        skip=skip,
                        )
            ]
            final = [
                {
                    "name": attributes.get("name"),
                    "version": attributes.get("version"),
                    "latest": str(getattr(package, "latest_version")),
                    "summary": attributes.get("summary"),
                    "home-page": attributes.get("home-page"),
                }
                for package in get_list.iter_packages_latest_infos(packages, options)
                for attributes in search_packages_info([package.key])
            ]
        else:
            # Pip 21.2 onwards use _ProcessedDists class to contain all the metadata of
            # a package instead of a dictionary.
            packages = [
                cast("_DistWithLatestInfo", d)
                for d in get_environment(options.path).iter_installed_distributions(
                    local_only=options.local,
                    user_only=options.user,
                    editables_only=options.editable,
                    include_editables=options.include_editable,
                    skip=skip,
                )
            ]
            final = [
                {
                    "name": str(getattr(attributes, "name")),
                    "version": str(getattr(attributes, "version")),
                    "latest": str(getattr(package, "latest_version")),
                    "summary": str(getattr(attributes, "summary")),
                    "home-page": str(getattr(attributes, "homepage")),
                }
                for package in get_list.iter_packages_latest_infos(packages, options)
                for attributes in search_packages_info([package.canonical_name])
            ]

        self.write_json(final, test_call)
        return final

    def install_package(self, _packages, params="", *args, **kwargs):
        # type: (str, str, tuple, dict) -> int
        """
        Installs the string of packages and returns the success code.
        """
        if compare_version((19, 1, 0), pip_version):
            install = InstallCommand()
        else:
            install = InstallCommand("Pippel", "Backend server for the Pippel service.")

        assert _packages, "`_packages` should not be an empty string."

        if "test" in params:
            test_call = True
            params = params.replace("test", "")
        else:
            test_call = False

        packages = _packages.split()
        if self.in_virtual_env():
            package_args = [sys.executable, "-m", "pip", "install", *packages,
                            "--upgrade"]
            self.write_out("Installing %s in a virtual environment\n" %
                           ", ".join(packages),
                           test_call)
        elif params:
            package_args = [sys.executable, "-m", "pip", "install", *packages,
                            "--upgrade", "--target", params]
            self.write_out("Installing %s to %s\n" %
                           (", ".join(packages), params),
                           test_call)
        else:
            package_args = [sys.executable, "-m", "pip", "install", *packages,
                            "--upgrade", "--user"]
            self.write_out("Installting %s in the local user space\n" %
                           packages,
                           test_call)

        with open(os.devnull, "w") as devnull:
            return subprocess.check_call(package_args, stdout=devnull)

    def remove_package(self, _packages, params="", *args, **kwargs):
        # type: (str, str, tuple, dict) -> int
        """
        Removes the string of packages and returns the success code.
        """
        if compare_version((19, 1, 0), pip_version):
            uninstall = UninstallCommand()
        else:
            uninstall = UninstallCommand(
                "Pippel", "Backend server for the Pippel service."
            )

        assert _packages, "`packages` should not be an empty string."

        if "test" in params:
            test_call = True
            params = params.replace("test", "")
        else:
            test_call = False

        packages = _packages.split()
        package_args = [sys.executable, "-m", "pip", "uninstall", *packages,
                        "--yes"]
        self.write_out("Uninstalling %s\n" % ", ".join(packages), test_call)

        with open(os.devnull, "w") as devnull:
            return subprocess.check_call(package_args, stdout=devnull)


if __name__ == "__main__":
    stdin = sys.stdin
    stdout = sys.stdout
    with open(os.devnull, "w") as null:
        sys.stdout = sys.stderr = null
    stdout.flush()
    PipBackend(stdin, stdout).serve_forever()


src/pippel.el
;;; pippel.el --- Frontend to python package manager pip -*- lexical-binding: t -*-

;; Copyright (C) 2017  Fritz Stelzer <brotzeitmacher@gmail.com>

;; Author: Fritz Stelzer <brotzeitmacher@gmail.com>
;; Maintainer: Arif Er <arifer612@protonmail.me>
;; URL: https://github.com/arifer612/pippel
;; Version: 1.4
;; Package-Requires: ((emacs "25.1") (s "1.11.0") (dash "2.12.0"))

;;; License:
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 3
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;;; CHANGELOG
;; v1.4 - 2021-12-05
;;   + Fix for pip >21.1.3 (#c3f10e6)
;;   + Formmating using black and add docstrings (#0e62405)
;; v1.3 - 2021-06-15
;;   + Fix for pip >19.1.1 (#2480fd3)
;;   + Maintainer changed to arifer612

;;; Code:

(require 'python)
(require 'tabulated-list)
(require 'json)
(require 's)
(require 'sort)
(require 'dash)

(defgroup pippel nil
  "Manager for pip packages."
  :prefix "pippel-"
  :group 'applications)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Customization Variables

(defcustom pippel-column-width-package 15
  "Width of the Package column."
  :type 'integer
  :group 'pippel)

(defcustom pippel-column-width-version 10
  "Width of the Version and Latest columns."
  :type 'integer
  :group 'pippel)

(defcustom pippel-menu-latest-face "orange"
  "Face for latest version when newer than installed version."
  :type 'face
  :group 'pippel)

(defcustom pippel-python-command "python"
  "Used Python interpreter."
  :type '(choice (const :tag "python" "python")
                 (const :tag "python2" "python2")
                 (const :tag "python3" "python3")
                 (string :tag "Other"))
  :group 'pippel)

(defcustom pippel-package-path (file-name-directory (locate-library "pippel"))
  "Directory for pippel.py.

If this is nil, it's assumed pippel can be found in the standard path."
  :type 'directory
  :group 'pippel)

(defcustom pippel-display-status-reporter t
  "Display progress-reporter."
  :type 'boolean
  :group 'pippel)

(defcustom pippel-buffer-display-method 'display-buffer
  "Default action when `pippel-list-packages' finished."
  :type '(choice (const :tag "Pop to buffer." 'display-buffer)
                 (const :tag "Pop to buffer." 'pop-to-buffer)
                 (const :tag "Don't display menu mode buffer." nil))
  :group 'pippel)

(defcustom pippel-debugging t
  "Turn on debugging for Pippel."
  :type 'boolean
  :group 'pippel)

;;;;;;;;;;;;;;;;;;;;;;;;
;;; Package menu mode

(defvar pippel-package-menu-mode-map
  (let ((map (make-sparse-keymap)))
    (set-keymap-parent map tabulated-list-mode-map)
    (define-key map (kbd "m") 'pippel-menu-mark-unmark)
    (define-key map (kbd "d") 'pippel-menu-mark-delete)
    (define-key map (kbd "U") 'pippel-menu-mark-all-upgrades)
    (define-key map (kbd "u") 'pippel-menu-mark-upgrade)
    (define-key map (kbd "r") 'pippel-list-packages)
    (define-key map (kbd "i") 'pippel-install-package)
    (define-key map (kbd "x") 'pippel-menu-execute)
    (define-key map (kbd "RET") 'pippel-menu-visit-homepage)
    (define-key map (kbd "q") 'quit-window)
    map)
  "Local keymap for `pippel-package-menu-mode' buffers.")

(define-derived-mode pippel-package-menu-mode tabulated-list-mode "Package Menu"
  "Major mode for browsing a list of installed pip packages."
  (setq buffer-read-only nil)
  (setq truncate-lines t)
  (setq tabulated-list-format
        `[("Package" ,pippel-column-width-package nil)
          ("Version" ,pippel-column-width-version nil)
          ("Latest" ,pippel-column-width-version nil)
          ("Description" 0 nil)])
  (setq tabulated-list-padding 2)
  (tabulated-list-init-header))

(defun pippel-menu-entry (pkg)
  "Return a package entry of PKG suitable for `tabulated-list-entries'."
  (let ((name (alist-get 'name pkg))
        (version (alist-get 'version pkg))
        (latest (alist-get 'latest pkg))
        (description (alist-get 'summary pkg))
        (home-page (alist-get 'home-page pkg)))
    (list name `[,(progn
                    (put-text-property 0 (length name) 'link home-page name)
                    name)
                 ,version
                 ,(if (or (string-match "[A-Za-z]+" version)
                          (string-match "[A-Za-z]+" latest))
                      (if (string= version latest)
                          latest
                        (propertize latest 'font-lock-face `(:foreground ,pippel-menu-latest-face)))
                    (if (version< version latest)
                        (propertize latest 'font-lock-face `(:foreground ,pippel-menu-latest-face))
                      latest))
                 ,description])))

(defun pippel-menu-generate (packages)
  "Re-populate the `tabulated-list-entries' with PACKAGES."
  (let ((buf (get-buffer-create
              (concat (when python-shell-virtualenv-root
                        (nth 1 (reverse (split-string python-shell-virtualenv-root "\\/"))))
                      "*Pip-Packages*"))))
    (with-current-buffer buf
      (pippel-package-menu-mode)
      (erase-buffer)
      (goto-char (point-min))
      (setq tabulated-list-entries
            (mapcar #'pippel-menu-entry (car packages)))
      (tabulated-list-print t)
      (let ((sort-fold-case t))
        (sort-lines nil (point-min) (point-max))))
    (when pippel-buffer-display-method
      (funcall pippel-buffer-display-method buf))))

;;;;;;;;;;;;;
;;; Server

(defvar pippel-process-name "pip-process"
  "Process name for pippel processes.")

(defvar pippel-process-buffer "*pip-process-buffer*"
  "Buffer name for pippel process buffers.")

(defvar pippel-debug-buffer "*pip-debug-buffer*"
  "Buffer name for pippel debug buffers.")

(defun pippel-running-p ()
  "Is pippel process running."
  (interactive)
  (or (get-buffer pippel-process-buffer)
      (process-live-p pippel-process-name)))

(defun pippel-open-process ()
  "Start and return pip process."
  (let ((buf pippel-process-buffer)
        (file (expand-file-name "pippel.py"
                                pippel-package-path)))
    (unless (file-exists-p file)
      (user-error "Can't find pippel in pippel-package-path"))
    (start-process pippel-process-name
                   buf
                   pippel-python-command
                   file)
    (let ((proc (get-buffer-process buf)))
      (set-process-filter proc 'pippel-process-filter)
      (set-process-sentinel proc 'pippel-process-sentinel)
      (accept-process-output proc 0.1)
      proc)))

(defun pippel-process-sentinel (proc _output)
  "The sentinel for pip-process."
  (with-current-buffer (process-buffer proc)
    (let ((objects nil)
          (json-array-type 'list))
      (goto-char (point-min))
      (while (not (eobp))
        (when (memq (char-after) '(?\{ ?\[))
          (push (json-read) objects))
        (forward-line))
      (when objects
        (pippel-menu-generate objects)))
    (while (process-live-p proc)
      (sleep-for 0.01))
    (kill-buffer (process-buffer proc)))
  (when pippel-display-status-reporter
    (remove-hook 'post-command-hook 'pippel-status-reporter)))

(defun pippel-process-filter (proc output)
  "Filter for pip-process."
  (let ((buf (process-buffer proc)))
    (with-current-buffer buf
      (insert output)
      (goto-char (point-max))
      (cond
       ((looking-back "Pip finished\n" nil)
        (message "Pip finished")
        (kill-process proc))
       ((looking-back "PIPPEL_ERROR >>>\n" nil)
        (if pippel-debugging
            (progn
              (message "Pippel error. Check the debug buffer for the error
message and please report it on GitHub.")
              (copy-to-buffer pippel-debug-buffer (point-min) (point-max)))
          (message "Pippel error. Set `pippel-debugging' to 't and please report
it as an issue on GitHub."))
        (kill-process proc))))))

(defun pippel-call-pip-process (proc command &optional arg1 arg2)
  "Send request to pip process."
  (process-send-string proc (concat (json-encode `((method . ,command)
                                                   (arg1 . ,arg1)
                                                   (arg2 . ,arg2)))
                                    "\n"))
  (when pippel-display-status-reporter
   (add-hook 'post-command-hook 'pippel-status-reporter)))

(defun pippel-status-reporter ()
  "Status indicator is shown in the echo area while pip process alive."
  (unless (active-minibuffer-window)
    (let ((progress-reporter (make-progress-reporter "Pip processing...")))
      (dotimes (i 1000)
        (when (pippel-running-p)
          (progress-reporter-update progress-reporter i)
          (sit-for 0.1))))))

;;;;;;;;;;;;;;;;;;
;;; Interaction

(defun pippel-remove-package (packages)
  "Uninstall provided PACKAGES."
  (pippel-call-pip-process (pippel-open-process) "remove_package" packages))

(defun pippel-upgrade-package (packages)
  "Update provided PACKAGES."
  (pippel-call-pip-process (pippel-open-process) "install_package" packages))

(defun pippel-menu-mark-unmark ()
  "Clear any marks on a package."
  (interactive)
  (tabulated-list-put-tag " " t))

(defun pippel-menu-mark-upgrade ()
  "Mark an upgradable package."
  (interactive)
  (unless (string= (aref (tabulated-list-get-entry) 1) (aref (tabulated-list-get-entry) 2))
    (tabulated-list-put-tag "U" t)))

(defun pippel-menu-mark-delete ()
  "Mark a package for deletion and move to the next line."
  (interactive)
  (tabulated-list-put-tag "D" t))

(defun pippel-menu-mark-all-upgrades ()
  "Mark all upgradable packages in the Package Menu."
  (interactive)
  (save-excursion
    (goto-char (point-min))
    (while (not (eobp))
      (unless (string= (aref (tabulated-list-get-entry) 1)
                       (aref (tabulated-list-get-entry) 2))
        (tabulated-list-put-tag "U" t))
      (forward-line))))

(defun pippel-menu-visit-homepage ()
  "Follow link provided by pip."
  (interactive)
  (save-excursion
    (beginning-of-line-text)
    (browse-url (get-text-property (point) 'link))))

(defun pippel-menu-execute ()
  "Perform marked Package Menu actions."
  (interactive)
  (let (upgrade-list
        delete-list
        cmd
        pkg-desc)
    (save-excursion
      (goto-char (point-min))
      (while (not (eobp))
        (setq cmd (char-after))
        (setq pkg-desc (tabulated-list-get-id))
        (cond ((eq cmd ?D)
               (push (substring-no-properties pkg-desc) delete-list))
              ((eq cmd ?U)
               (push (substring-no-properties pkg-desc) upgrade-list)))
        (forward-line)))
    (unless (or delete-list upgrade-list)
      (user-error "No operations specified"))
    (let* ((del (when delete-list
                  (concat (format "Delete %d package%s "
                                  (length delete-list)
                                  (if (> (length delete-list) 1)
                                      "s" ""))
                          (replace-regexp-in-string " " ", "
                                                    (format "%s" delete-list)))))
           (up (when upgrade-list
                 (concat (format "Upgrade %d package%s "
                                 (length upgrade-list)
                                 (if (> (length upgrade-list) 1)
                                     "s" ""))
                         (replace-regexp-in-string " " ", "
                                                   (format "%s" upgrade-list)))))
           (msg (if (and del up)
                    (concat del " and " up)
                  (or del up))))
      (when (yes-or-no-p (format "%s" msg))
        (when upgrade-list
          (pippel-upgrade-package (mapconcat 'identity upgrade-list " ")))
        (while (pippel-running-p)
          (sleep-for 0.01))
        (when delete-list
          (pippel-remove-package (mapconcat 'identity delete-list " ")))
        (while (pippel-running-p)
          (sleep-for 0.01))
        (pippel-list-packages)))))

;;;###autoload
(defun pippel-install-package (arg)
  "Prompt user for a string containing packages to be installed."
  (interactive "P")
  (let ((pkg (read-from-minibuffer "Enter package name: "))
        (proc (pippel-open-process))
        (dir (--if-let arg
                 (read-file-name "Directory: "))))
    (pippel-call-pip-process proc "install_package" (s-trim pkg) dir)
    (when (string= major-mode "pippel-package-menu-mode")
      (while (pippel-running-p)
        (sleep-for 0.01))
      (pippel-list-packages))))

;;;###autoload
(defun pippel-list-packages ()
  "Display a list of installed packages."
  (interactive)
  (pippel-call-pip-process (pippel-open-process) "get_installed_packages"))

;;;###autoload
(defun pippel-list-user-packages ()
  "Display a list of user-installed system packages."
  (interactive)
  (pippel-call-pip-process (pippel-open-process) "get_installed_packages"
                           "--user"))

(provide 'pippel)
;;; pippel.el ends here
